docarray/utils/reduce.py

```diff
@@ -0,0 +1,196 @@
+from docarray import DocumentArray
+from typing import List, Optional, Dict, _GenericAlias  # type: ignore
+from docarray.base_document import BaseDocument
+
+from collections import namedtuple
+
+# Declaring namedtuple()
+_FieldGroups = namedtuple(
+    '_FieldGroups',
+    [
+        'simple_non_empty_fields',
+        'list_fields',
+        'set_fields',
+        'dict_fields',
+        'nested_docarray_fields',
+        'nested_docs_fields',
+    ],
+)
+
+
+def _group_fields(doc: 'BaseDocument') -> _FieldGroups:
+    simple_non_empty_fields: List[str] = []
+    list_fields: List[str] = []
+    set_fields: List[str] = []
+    dict_fields: List[str] = []
+    nested_docs_fields: List[str] = []
+    nested_docarray_fields: List[str] = []
+
+    for field_name, field in doc.__fields__.items():
+        field_type = doc._get_field_type(field_name)
+        if not isinstance(field_type, _GenericAlias) and issubclass(
+            field_type, DocumentArray
+        ):
+            nested_docarray_fields.append(field_name)
+        elif isinstance(field_type, _GenericAlias) and field_type.__origin__ is list:
+            list_fields.append(field_name)
+        elif isinstance(field_type, _GenericAlias) and field_type.__origin__ is set:
+            set_fields.append(field_name)
+        elif isinstance(field_type, _GenericAlias) and field_type.__origin__ is dict:
+            dict_fields.append(field_name)
+        v = getattr(doc, field_name)
+        if v:
+            if isinstance(v, BaseDocument):
+                nested_docs_fields.append(field_name)
+            else:
+                simple_non_empty_fields.append(field_name)
+    return _FieldGroups(
+        simple_non_empty_fields,
+        list_fields,
+        set_fields,
+        dict_fields,
+        nested_docarray_fields,
+        nested_docs_fields,
+    )
+
+
+def reduce_docs(doc1: 'BaseDocument', doc2: 'BaseDocument') -> 'BaseDocument':
+    """
+    Reduces doc1 and doc2 into one Document in-place. Changes are applied to doc1.
+    Reducing 2 Documents consists in the following:
+     - setting data properties of the second Document to the first Document
+     if they are empty (priority to the left-most Document)
+     - Concatenating lists and updating sets
+     - Reducing recursively Documents and DocumentArrays
+     - Updating Dictionaries of the left with the right
+    :param doc1: first Document to be reduced. Change is applied in-place
+    :param doc2: second Document to be reduced
+    :return: The reduced Document
+    """
+    doc1_fields = _group_fields(doc1)
+    doc1_simple_non_empty_fields = doc1_fields.simple_non_empty_fields
```

Comment:
do u need all of this item assignment ? I think you can just use `doc1_fields.simple_non_empty_fields` everywhere where instead of `doc1_simple_non_empty_fields`

Reply:
will do